File: classes/web/json/json_data.py
import json


class Fund:
    def __init__(self,name,desc, amount,return_type,return_rate,default_rate, mimic_calendar_year, wt_age):
        self.name=name
        self.amount=amount
        self.desc=desc
        self.return_type = return_type #Flat, Index
        self.return_rate=return_rate
        self.default_rate = default_rate
        self.mimic_calendar_year = mimic_calendar_year
        self.wt_age=wt_age
    
    def to_dict(self):
        return {'name': self.name,'desc': self.desc, 'amount': self.amount, 'return_type': self.return_type, 
                'return_rate': self.return_rate,'default_rate': self.default_rate,'mimic_calendar_year': self.mimic_calendar_year, 
                'wt_age': self.wt_yr}    
    
    @classmethod
    def from_dict(cls, fund_dict):
        return cls(fund_dict['name'], fund_dict['desc'], fund_dict['amount'], fund_dict['return_type'], fund_dict['return_rate'],
                    fund_dict['default_rate'] , fund_dict['mimic_calendar_year'], fund_dict['wt_age'])
        
class Expense:
    
    def __init__(self,name,min_rate,inflation):
        self.name=name
        self.min_rate=min_rate
        self.inflation=inflation
        self.amounts=[]
    
    def add_amounts(self, detail):
        self.amounts.append(detail)

    def to_dict(self):
        return {'name': self.name, 'min_rate': self.min_rate, 'inflation': self.inflation, 
                           'amounts': [detail.to_dict() for detail in self.amounts]}

    @classmethod
    def from_dict(cls, expense_dict):
        expense = cls(expense_dict['name'], expense_dict['min_rate'],expense_dict['inflation'])
        for amounts_dict in expense_dict['amounts']:
            expense.add_amounts(Expense_detail.from_dict(amounts_dict))
        return expense
    
class Expense_detail:
    def __init__(self,amount,start_age,until_age,occur_yr, tied_to_age):
        self.amount=amount
        self.start_age=start_age
        self.until_age=until_age
        self.occur_yr=occur_yr
        self.tied_to_age=tied_to_age
    def to_dict(self):
        return {
            'amount': self.amount,
            'start_age': self.start_age,
            'until_age': self.until_age,
            'occur_yr': self.occur_yr,
            'tied_to_age': self.tied_to_age
        }

    @classmethod
    def from_dict(cls, fund_dict):
        return cls(fund_dict['amount'], fund_dict['start_age'], fund_dict['until_age'],fund_dict['occur_yr'],fund_dict['tied_to_age'])

    
class Income:

    def __init__(self,name,amount,dep_to_fund):
        self.name=name
        self.amount=amount
        self.dep_to_fund=dep_to_fund
        self.details:list[Income_detail]=[]

    def add_detail(self, detail):
        self.details.append(detail)
    
    @classmethod
    def from_dict(cls, income_dict):
        income = cls(income_dict['name'], income_dict['amount'],income_dict['dep_to_fund'])
        for detail_dict in income_dict['growths']:
            income.add_detail(Income_detail.from_dict(detail_dict))
        return income

    def to_dict(self):
        return {'name': self.name, 'amount': self.amount, 'dep_to_fund': self.dep_to_fund, 
                           'growths': [detail.to_dict() for detail in self.details]}
            
class Income_detail:
    def __init__(self,growth_rate,start_age,until_age):
        self.growth_rate=growth_rate
        self.start_age=start_age
        self.until_age=until_age
    
    def to_dict(self):
        return {
            'growth_rate': self.growth_rate,
            'start_age': self.start_age,
            'until_age': self.until_age
        }

    @classmethod
    def from_dict(cls, fund_dict):
        return cls(fund_dict['growth_rate'], fund_dict['start_age'], fund_dict['until_age'])
    

class Strategic:
    
    #def __init__(self,apply_min,apply_bucket,risk_fund,safer_fund,rebal_pause_option,retire_calendar_year,default_return_rate,expected_return_rate):
    def __init__(self,apply_min,apply_bucket,risk_fund,safer_fund,rebal_pause_option,expected_return_rate):
        
        self.apply_min=apply_min
        self.apply_bucket=apply_bucket
        self.risk_fund=risk_fund
        self.safer_fund=safer_fund
        #self.retire_calendar_year=retire_calendar_year
        self.rebal_pause_option=rebal_pause_option
        #self.default_return_rate=default_return_rate
        self.expected_return_rate=expected_return_rate
        self.details:list[Strategic_detail]=[]

    def add_detail(self, detail):
        self.details.append(detail)

    @classmethod
    def from_dict(cls, strategic_dict):
        strategic = cls(strategic_dict['apply_min'], strategic_dict['apply_bucket'],strategic_dict['risk_fund'],
                       strategic_dict['safer_fund'],strategic_dict['rebal_pause_option']
                       #,strategic_dict['retire_year'],strategic_dict['default_return_rate']
                       ,strategic_dict['expected_return_rate'])
        for detail_dict in strategic_dict['rebals']:
            strategic.add_detail(Strategic_detail.from_dict(detail_dict))
        return strategic
    
    def to_dict(self):
        return {'min_rate': self.min_rate, 'apply_min': self.apply_min, 'risk_fund': self.risk_fund, 
                            'safer_fund':self.safer_fund,'rebal_pause_option':self.rebal_pause_option,
                            'retire_year':self.retire_calendar_year,'default_return_rate':self.default_return_rate,'expected_return_rate':self.expected_return_rate,
                           'rebals': [detail.to_dict() for detail in self.details]}
class Strategic_detail:
    def __init__(self,risk_fund_ratio,safer_fund_ratio,start_age,until_age,occur_yr):
        self.risk_fund_ratio=risk_fund_ratio
        self.safer_fund_ratio=safer_fund_ratio
        self.start_age=start_age
        self.until_age=until_age
        self.occur_yr=occur_yr
    def to_dict(self):
        return {
            'risk_fund_ratio': self.risk_fund_ratio,
            'safer_fund_ratio': self.safer_fund_ratio,
            'start_age': self.start_age,
            'until_age': self.until_age,
            'occur_yr': self.occur_yr,
        }
    @classmethod
    def from_dict(cls, detail_dict):
        return cls(detail_dict['risk_fund_ratio'], detail_dict['safer_fund_ratio'],
                   detail_dict['start_age'], detail_dict['until_age'], detail_dict['occur_yr'])

class Plan_Data:    
    def __init__(self,current_age,retire_age):
        self.current_age=int(current_age)
        self.retire_age=int(retire_age)
        self.incomes=[]
        self.funds=[]
        self.expenses=[]
        self.strategic=None
    
    def add_expense(self, expense):
        self.expenses.append(expense)
    def add_fund(self, fund):
        self.funds.append(fund)
    def add_income(self, income):
        self.incomes.append(income)
    def set_strategic(self,strategic):
        self.strategic=strategic

    def to_json(self):
        return json.dumps({ 'current_age':self.current_age, 'retire_age':self.retire_age, 
                            'expenses': [expense.to_dict() for expense in self.expenses],
                            'incomes': [income.to_dict() for income in self.incomes],
                            'funds': [fund.to_dict() for fund in self.funds],
                            'strategic':self.strategic.to_dict()
                            }, indent=4)
    @classmethod
    def from_json(cls, json_string):
        plan_data_dict = json.loads(json_string)
        plan_data = cls( plan_data_dict['current_age'], plan_data_dict['retire_age'])
        for expenses_dict in plan_data_dict['expenses']:
            plan_data.add_expense(Expense.from_dict(expenses_dict))
        if 'incomes' in plan_data_dict:
            for incomes_dict in plan_data_dict['incomes']:
                plan_data.add_income(Income.from_dict(incomes_dict))
        for fund_dict in plan_data_dict['funds']:
            plan_data.add_fund(Fund.from_dict(fund_dict))
        plan_data.strategic=Strategic.from_dict(plan_data_dict['strategic'])    
        return plan_data

File: static/js/utilities/json_data.js
class Fund {
  constructor(name, desc, amount, return_type, return_rate, default_rate, mimic_calendar_year, wt_age) {
      this.name = name;
      this.amount = toFloat(amount);
      this.desc = desc;
      this.return_type = return_type;
      this.return_rate = return_rate;
      this.default_rate = default_rate;
      this.mimic_calendar_year = mimic_calendar_year;
      this.wt_age = wt_age;
  }
}

class Expense {
  constructor(name, min_rate, inflation) {
      this.name = name;
      this.min_rate = toFloat(min_rate);
      this.inflation = toFloat(inflation);
      this.amounts = [];
  }

  addAmounts(amount) {
      this.amounts.push(toFloat(amount));
  }
  addAmountsFirstPos(amount){
    this.amounts.unshift(toFloat(amount));
  }
}

class ExpenseDetail {
  constructor(amount, start_age, until_age, occur_yr, tied_to_age) {
    this.amount = toFloat(amount);
    this.start_age = toInt(start_age);
    this.until_age = toInt(until_age);
    this.occur_yr = toInt(occur_yr);
    this.tied_to_age = tied_to_age
  }
}

class Income {
  constructor(name, amount, dep_to_fund) {
      this.name = name;
      this.amount = toFloat(amount);
      this.dep_to_fund = dep_to_fund;
      this.growths = [];
  }

  addGrowths(growth) {
      this.growths.push(toFloat(growth));
  }
  addGrowthsFirstPos(growth){
      this.growths.unshift(toFloat(growth));
  }
}

class IncomeDetail {
  constructor(growth_rate, start_age, until_age) {
      this.growth_rate = toFloat(growth_rate);
      this.start_age = toInt(start_age);
      this.until_age = toInt(until_age);
  }
}

class Strategic {
  //constructor(apply_min, apply_bucket, risk_fund, safer_fund,rebal_pause_option, retire_year,default_return_rate,expected_return_rate)
  constructor(apply_min, apply_bucket, risk_fund, safer_fund,rebal_pause_option, expected_return_rate) {
      this.apply_min = apply_min;
      this.apply_bucket = apply_bucket;
      this.risk_fund = risk_fund;
      this.safer_fund = safer_fund;
      this.rebal_pause_option = rebal_pause_option;
      //this.retire_year = toInt(retire_year);
      //this.default_return_rate = toFloat(default_return_rate);
      this.expected_return_rate = toFloat(expected_return_rate);
      this.rebals = [];
  }

  addRebals(detail) {
      this.rebals.push(detail);
  }

  addRebalsFirstPos(detail){
    this.rebals.unshift(detail);
  }
}


class StrategicDetail{
    constructor(risk_fund_ratio, safer_fund_ratio,start_age,until_age,occur_yr){
      this.risk_fund_ratio=toFloat(risk_fund_ratio)
      this.safer_fund_ratio=toFloat(safer_fund_ratio)
      this.start_age=toInt(start_age)
      this.until_age=toInt(until_age)
      this.occur_yr=toInt(occur_yr)
    }
}

class Plan_Data{
    constructor(current_age,retire_age){
          this.current_age=current_age
          this.retire_age=retire_age
          this.incomes=[]
          this.funds=[]
          this.expenses=[]
          this.strategic=null
    }
      
    add_expense(expense){
          this.expenses.push(expense)
    }
    add_fund(fund){
      this.funds.push(fund)
    }
    add_income(income){
      this.incomes.push(income)
    }
    set_strategic(strategic){
      this.strategic=strategic
    }
    addExpenseFirstPos(expense){
      this.expenses.unshift(expense);
    }
    addFundFirstPos(fund){
      this.funds.unshift(fund);
    }
    addIncomeFirstPos(income){
      this.incomes.unshift(income);
    }
}
function toInt(inputValue) {
  if (typeof inputValue === 'string'){
    if(! (inputValue !== null && inputValue !== undefined && inputValue !== ''))
        return 0
    str = inputValue.replace(/\s+/g, ''); // remove spaces
    let num = parseInt(str); // convert to integer
    return num
  }
  return inputValue
}
function toFloat(inputValue) {
  if (typeof inputValue === 'string'){
    if(! (inputValue !== null && inputValue !== undefined && inputValue !== ''))
        return 0
    str = inputValue.replace(/\s+/g, ''); // remove spaces
    let num = parseFloat(str); // convert to integer
    return num
  }
  return inputValue
}

/** 
e1=new Expense("E1",0.8,5);
e1.addAmounts(new ExpenseDetail(10,1,10,1));
e1.addAmounts(new ExpenseDetail(100,11,20,1));
e1.addAmounts(new ExpenseDetail(500,21,30,1));

e2=new Expense("E2",0.8,5);
e2.addAmounts(new ExpenseDetail(10,1,10,1));
e2.addAmounts(new ExpenseDetail(100,11,20,1));
e2.addAmounts(new ExpenseDetail(500,21,30,1));


i1= new Income("I1",1000,"EPF");
i1.addGrowths(new IncomeDetail(3.0, 5,9));
i1.addGrowths(new IncomeDetail(5.0, 10,19));
i1.addGrowths(new IncomeDetail(6.0, 20,29));
i2= new Income("I1",1000,"Stocks");
i2.addGrowths(new IncomeDetail(3.0, 5,9));
i2.addGrowths(new IncomeDetail(5.0, 10,19));
i2.addGrowths(new IncomeDetail(6.0, 20,29));


s1=new Strategic(true,true,"E","F","A", 1998);
s1.addRebals(new StrategicDetail(30,40,1,100,2))
s1.addRebals(new StrategicDetail(30,40,1,100,2))


const fund1= new Fund('EPF', 't', 1000,5.0,55);
const fund2= new Fund('EPF','t2', 1000,5.0,55);

p = new Plan_Data(23,2008)
p.set_strategic(s1)
p.add_expense(e1)
p.add_expense(e2)
p.add_income(i1)
p.add_income(i2)
p.add_fund(fund1)
p.add_fund(fund2)

const prettyJSON = JSON.stringify(p, null, 2);
console.log(prettyJSON);
*/
 /** 
console.log("=================READING=====================");

// Assuming your JSON file is named "data.json"
const fs = require('fs');
const path = require('path');

const jsonFilePath = path.join(__dirname, 'json_data_py.json');
const rawJsonString = fs.readFileSync(jsonFilePath, 'utf8');
// Use the raw JSON string here
//console.log(rawJsonString);
p2 = JSON.parse(rawJsonString)
console.log("after parse");

p2.expenses.forEach(expense => {
  console.log(`Expense: ${expense.name}`);
  expense.amounts.forEach(amount => {
      console.log(`  Amount: ${amount.amount}, Year From: ${amount.start_yr}, Year To: ${amount.until_yr}`);
    });
  });
  */